src/qt/textctrl.cpp

```diff
@@ -14,19 +14,364 @@
 #include "wx/qt/private/winevent.h"
 #include "wx/qt/private/utils.h"
 
+#include <algorithm>
 #include <QtWidgets/QLineEdit>
 #include <QtWidgets/QTextEdit>
 
+
+namespace
+{
 class wxQtLineEdit : public wxQtEventSignalHandler< QLineEdit, wxTextCtrl >
 {
 public:
     wxQtLineEdit( wxWindow *parent, wxTextCtrl *handler );
 
 private:
-    void textChanged(const QString &text);
+    void textChanged();
     void returnPressed();
 };
 
+class wxQtTextEdit : public wxQtEventSignalHandler< QTextEdit, wxTextCtrl >
+{
+public:
+    wxQtTextEdit( wxWindow *parent, wxTextCtrl *handler );
+
+private:
+    void textChanged();
+};
+
+class wxQtMultiLineEdit : public wxQtEdit
+{
+public:
+    wxQtMultiLineEdit(QTextEdit *edit) : edit(edit)
+    {
+    }
+
+    virtual bool IsModified() const wxOVERRIDE
+    {
+        return edit->isWindowModified();
+    }
+
+    virtual wxString DoGetValue() const wxOVERRIDE
+    {
+        return wxQtConvertString( edit->toPlainText() );
+    }
+
+    virtual long GetInsertionPoint() const wxOVERRIDE
+    {
+        QTextCursor cursor = edit->textCursor();
+        return cursor.anchor();
+    }
+
+    virtual QWidget *GetHandle() const wxOVERRIDE
+    {
+        return edit;
+    }
+
+    virtual int GetNumberOfLines() const wxOVERRIDE
+    {
+        const wxString &value = DoGetValue();
+        return std::count(value.begin(), value.end(), '\n') + 1;
+    }
+
+    std::pair<size_t, size_t> getLineStart(long lineNo, const wxString &value) const
```

Comment:
I'd strongly prefer to avoid the use of `std::pair` as it makes the code unreadable. Having a (private) `struct wxLineInfo { long startPos, endPos; }` would be much better.

Reply:
Replaced with a struct